File: src/NodeParser/ObjectTypeNodeParser.ts
import ts from "typescript";
import type { Context } from "../NodeParser.js";
import type { SubNodeParser } from "../SubNodeParser.js";
import type { BaseType } from "../Type/BaseType.js";
import { ObjectType } from "../Type/ObjectType.js";
import { getKey } from "../Utils/nodeKey.js";

export class ObjectTypeNodeParser implements SubNodeParser {
    public supportsNode(node: ts.KeywordTypeNode): boolean {
        return node.kind === ts.SyntaxKind.ObjectKeyword;
    }

    public createType(node: ts.KeywordTypeNode, context: Context): BaseType {
        return new ObjectType(`object-${getKey(node, context)}`, [], [], true, true);
    }
}

File: internal/parser/object_type_node_parser.go
package parser

import (
	"github.com/microsoft/typescript-go/shim/ast"

	"github.com/vega/ts-json-schema-generator/internal/types"
)

// ObjectTypeNodeParser handles the `object` keyword
// (src/NodeParser/ObjectTypeNodeParser.ts).
type ObjectTypeNodeParser struct{}

func NewObjectTypeNodeParser() *ObjectTypeNodeParser {
	return &ObjectTypeNodeParser{}
}

func (p *ObjectTypeNodeParser) SupportsNode(node *ast.Node) bool {
	return node.Kind == ast.KindObjectKeyword
}

func (p *ObjectTypeNodeParser) CreateType(node *ast.Node, ctx *Context, _ *types.ReferenceType) types.Type {
	return types.NewObjectType("object-"+GetNodeKey(node, ctx), nil, nil, true, true)
}
